### PR TITLE
#55 - added DKIM support -i 55

### PR DESCRIPTION
**Pull request by bilak**
Added support for DKIM
Maybe something can be refactored, but it's working correctly now.

### CODE REVIEW

**Comment by tomav** [comment]
Thank you @bilak!

My questions:
- does it work with multi-domains because I see a `domainname` ENV added to the project.
- I'm not a DKIM expert but is there a way to add integration tests?

Thank you.

File: postfix/TrustedHosts [hunk 1]
```diff
@@ -0,0 +1,2 @@
+127.0.0.1
+localhost
```
Comment (tomav) [inline postfix/TrustedHosts]: Missing new line.

File: postfix/default-opendkim [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+#DAEMON_OPTS=""
+#
+# Uncomment to specify an alternate socket
+# Note that setting this will override any Socket value in opendkim.conf
+#SOCKET="local:/var/run/opendkim/opendkim.sock" # default
+#SOCKET="inet:54321" # listen on all interfaces on port 54321
+#SOCKET="inet:12345@localhost" # listen on loopback on port 12345
+#SOCKET="inet:12345@192.0.2.1" # listen on 192.0.2.1 on port 12345
+
+SOCKET="inet:12301@localhost"
```
Comment (tomav) [inline postfix/default-opendkim]: Missing new line.

File: postfix/opendkim.conf [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+KeyTable                refile:/etc/opendkim/KeyTable
+SigningTable            refile:/etc/opendkim/SigningTable
+
+Mode                    sv
+PidFile                 /var/run/opendkim/opendkim.pid
+SignatureAlgorithm      rsa-sha256
+
+UserID                  opendkim:opendkim
+
+Socket                  inet:12301@localhost
```
Comment (tomav) [inline postfix/opendkim.conf]: Missing new line.

File: start-mailserver.sh [hunk 4]
```diff
@@ -1,10 +1,55 @@
...
+  echo "Creating DKIM SigningTable"
+  echo "*@$domainname mail._domainkey.$domainname" > /etc/opendkim/SigningTable
+fi
+echo "Changing permissions on /etc/opendkim"
+# chown entire directory
+chown -R opendkim:opendkim /etc/opendkim/
+# And make sure permissions are right
+chmod -R 0700 /etc/opendkim/keys/
+
+# Opendkim:
```
Comment (tomav) [inline start-mailserver.sh]: These next lines are needed for debug but not on a production server.
(I have to clean the current debug by the way..)

**Comment by bilak** [comment]
- It's now working only for one domain. We can try to convert it to multi-domain in future, but there must be something with what we can decide there's more domains (maybe domainname values separated by comma)
- About integration tests - only thing which we can check is information in log `mail opendkim[579]: 359D4A1245: DKIM-Signature field added (s=mail, d=domain.com)`. I don't know much about postfix neither about docker so your help will be needed.

**Comment by tomav** [comment]
- You probably don't have to provide a domain list since it already exists and is parsed to create users, aliases and postfix configuration. https://github.com/tomav/docker-mailserver/blob/master/start-mailserver.sh#L20
- For the tests, I will do it, no problem.

**Comment by bilak** [comment]
- So if I understand it correctly, there could be function created from my script which will accept only domain name (think about how to check duplicated processing - for example ten users with same domain and another 10 users with another domain). There's need to create (if not exists) DKIM (privatet, txt), add entry to KeyTab and to SigningTable.

**Comment by tomav** [comment]
The `postfix/vhost` file is generated by startup script and contains unique domains.
(I just found a strange behaviour on my server and trying to fix it but it does the trick).
I think you can use this file as reference and paste your DKIM startup lines after its generation.

The point is, for these domains, enable DKIM if necessary information are provided.

**Comment by tomav** [comment]
(about the bug mentioned above, it was not the container but my `virtual` alias list, so no worry.)

**Comment by tomav** [comment]
Thank you @bilak

**Pull request merged**